src/Data/GADT/TH/Internal.hs

```diff
@@ -35,23 +37,39 @@ classHeadToParams t = (h, reverse reversedParams)
       _ -> (headOfType t, [])
 
 -- Do not export this type family, it must remain empty. It's used as a way to trick GHC into not unifying certain type variables.
-type family Skolem :: k -> k
+type family Skolem :: Data.Kind.Type -> k -> k
+
+-- Do not export this type. It's part of the skolem trick above.
+data Unknown = Unknown
 
 skolemize :: Set Name -> Type -> Type
 skolemize rigids t = case t of
   ForallT bndrs cxt t' -> ForallT bndrs cxt (skolemize (Set.difference rigids (Set.fromList (map tvName bndrs))) t')
   AppT t1 t2 -> AppT (skolemize rigids t1) (skolemize rigids t2)
   SigT t k -> SigT (skolemize rigids t) k
   VarT v -> if Set.member v rigids
-    then AppT (ConT ''Skolem) (VarT v)
+    then AppT (AppT (ConT ''Skolem) (ConT ''Unknown)) (VarT v)
```

Comment:
What is the `Unknown` argument for?

Reply:
It's an attempt at making the types harder for GHC to incorrectly unify, but actually, we should check if there's a version for which this actually fixed things. (I seem to recall there was at least one as we were experimenting, but maybe the results were still wrong enough that we ended up excluding it anyway.)

Reply:
I removed it. It's no longer necessary.